**1. ETL (ORIGINAL DATASET)**

# 1.1 Import the required libraries

The following libraries are common in Python programming environments, especially in the context of data analysis and data science. 


**Pandas (import pandas as pd):**

Pandas is a Python library that provides flexible date structures and date analysis tools. 
Importing it as pd is a common convention to shorten the library name and make the code more consice. 
Pandas is widely used to manipulate and analyze tabular datasets.


**JSON(import json):**

JSON (JavaScript Object Notation) is a lightweight, human-readable data interchange format.
The json library in Python allows for the serialization and deserialization of data in JSON format.
It can be used to read and write data in this format.


**AST (import ast):**

AST (Abstract Syntax Tree) is a hierarchical representation of the syntactic structure of a Python program.
The ast library allows you to analyze and manipulate the abstract syntax tree of a Python source code.
It can be useful for performing static analysis of the code.


**Regular Expressions (import re)**

The 're' module provides regular expression operations in Python.
Regular expressions are search patterns used to match text strings.
They are powerful tools for manipulating and searching specific patterns within text strings.


**%load_ext autoreload & %autoreload 2:**

These commands are specific to jupyter notebooks and are used to automatically reload modules before executing a cell '%load_ext autoreload' enables automatic module reloading, and '%autoreload 2' sets the autoreload to be more aggressive, even reloading module functions.


**Warnings (import warnings):**

The warnings module provides tools for controlling the warnings emitted by Python.
In this case, it's being configured to ignore warnings, which can be useful to prevent warnings from filling up the console output and distracting during code execution.
In summary, these imports are common in data analysis and date science environments in Python, providing tools for manipulating data, working with JSON, parsing regular expressions, and handling warnings.
Additionaly, the '%load_ext autoreload and %autoreload 2' commands are specific to jupyter notebooks and are used to facilitate interactive development.


**TextBlob (from textblob import TextBlob)**

TextBlob is a Python library that provides tools for natural language processing (NLP).
It enables tasks such as sentiment analysis, extraction of key phrases, part-of-speech tagging, etc.


**Nltk (import nltk):**

The Natural Language Toolkit (nltk) library is another powerful tool for natural language processing in Python.
It provides a range of modules and resources for tasks such as tokenization, syntactic analysis, stemming, among others.
In this case, you are importing the entire nltk module.


**CSV (import csv)**

The 'csv' module in Python provides funtionality for working with CSV (Comma-Separated Values) files.
CSV files are a common format for storing tabular data, where each row in the file represents a data entry and values are separated by commas or another delimiter.

The 'csv' module in Python provides functions for reading data from CSV files and writing data to CSV files.
Some important functions include 'csv.reader()' for reading a CSV file and 'csv.writer()' for writing to a CSV file.




In [123]:

'''Necessary libraries.'''
import pandas as pd                 # Pandas for tabular data manipulation.
import json                         # Module for working with JSON.
import ast                          # Module for evaluating Python literal expressions.
import re                           # Module for working with regular expressions.
from textblob import TextBlob       # I import TextBlob from the textblob library.
import nltk                         # Natural Language Toolkit.
import csv                          # I import the CSV module into Python.

'''Enable auto-reload of modules before executing a cell'''
%load_ext autoreload
%autoreload 2

'''Import the warning module and set it to ignore all warnings'''
import warnings
warnings.filterwarnings("ignore")

The autoreload extension is already loaded. To reload it, use:
  %reload_ext autoreload


# 1.2 Auxiliary Functions

**1.2.1 Check Data Type**

We're checking the data types contained in the columns of "df_games".

This function called 'check data types' takes a DataFrame as input and returns a new DataFrame that provides information about the data types and the quantity of null values in each column of the input DtaFrame.

This function offers a useful summary of the features of the columns in a DataFrame, incluiding data types and the quantity of null values.
It's especially useful for conducting an initial data quality analysis on a dataset.

In [124]:
def check_data_type(df):
    
    '''A dictionary (my_dict) is created with five keys: "field_name", "data_type", "non_null_%", "null_%" and "nulls". 
    These keys will be used to store information about each column of the DataFrame.'''
    
    my_dict = {"field_name" : [], "data_type" : [], "non_null_%" : [], "null_%" : [], "nulls" : []}
    
    
    '''A loop is performed over all the columns of the DataFrame df'''
    for column in df.columns:
        percentage_non_nulls = (df[column].count() / len(df) * 100)     # The percentage of non-null values in the current column is calculated 
        my_dict['field_name'].append(column)                            # The current column is added to the list under the key 'field_name'        
        my_dict['data_type'].append(df[column].apply(type).unique())    # Obtain the unique data type in the current column and add it to the list under the key 'data type'.
        my_dict['non_null_%'].append(round(percentage_non_nulls, 2))    # Add the percentage of non-null values to the list under the key 'non_null_%'.
        my_dict['null_%'].append(round(100 - percentage_non_nulls, 2))  # Add the percentage of null values to the list under the key 'null_%'.
        my_dict['nulls'].append(df[column].isnull().sum())              # Add the number of null values in the current column to the list under the key 'nulls'.
        
    '''The dictionari my_dict is used to create a new DataFrame called df_info.'''
    df_info = pd.DataFrame(my_dict)
        
    '''The function returns the DataFrame df_info containing information about each column,
    incluiding the column name, data type, percentage of non-null values, percentage of null values, and the number of null values.'''
    return df_info

**1.2.2 Check duplicates by columns**

The following function provides a useful tool for identifying and sorting duplicate rows in a pandas DataFrame based on the values of a specifict column.
In our case, it can be useful for data analysis when examining and handling duplicates based on a particular column.

In [125]:
def check_duplicates_by_columns(df, column):
    
    '''Duplicate rows are filtered'''
    duplicated_rows = df[df.duplicated(subset=column, keep=False)]
    if duplicated_rows.empty:
        return 'There are no duplicates'
    
    '''The duplicate rows are sorted for comparison'''
    duplicated_rows_sorted = duplicated_rows.sort_values(by=column)
    return duplicated_rows_sorted

**1.2.3 Convert date**

The following function takes a date string in a specifict format, attempts to extract and convert that date to a different format ('YYYY-MM-DD'),
and returns the resulting date or an error message if the string does not match the expected format.

In [126]:
def convert_date(date_string):
    
    '''Searches the date string for a pattern matching the format "month, day, year'''
    match = re.search(r'(\w+\s\d{1,2},\s\d{4})', date_string)

    if match:
        '''If there is a match, it extracts the date string'''
        date_str = match.group(1)
        try:
            '''It tries to convert the date string to a Pandas date object'''
            date_dt = pd.to_datetime(date_str)
            '''It formats the resulting date into the "YYY-MM-DD" format and retuns it'''
            return date_dt.strftime('%y-%m-%d')
        except:
            '''In case of an error during conversion, it returns "Invalid date".'''
            return 'Invalid date'
    
    else:
        '''If there is no match, it returns "Invalid format".'''
        return 'Invalid format'

**1.2.4 Sentiment analysis**

This finction provides a basic way to categorize the sentiment of a text into positive, negative, or neutral based on the polarity calculated by TextBlob.

In [127]:
'''Definition of the sentiment analysis function.'''
def sentiment_analysis(review):             # Checks if the review is None.
    
    if review is None:                      # If affirmative, it returns 1, which could be interpreted as a neutral.
        return 1
    
    analysis = TextBlob(review)             # Creates an instance of the TextBlob class with the provided review.
    polarity = analysis.sentiment.polarity  # Gets the sentiment polarity from the TextBlob analysis.
    
    if polarity < -0.2:                     # Compares the polarity with thresholds to determine the overall sentiment.
        return 0                            # If the polarity is less than -0.2, it's considered a negative sentiment and returns 0.
    
    elif polarity > 0.2:                    # If the polarity is greater than 0.2, it's considered a positive sentiment and returns 2.
        return 2
    
    else:
        return 1                            # In other cases, it return 1, which could be interpreted as a neutral sentiment.
    
             


**1.2.5 Analysis of example reviews by sentiment**

The function 'examples_reviwe_by_sentiment' is used to analyze and present examples of reviews classified according to their sentiments.
The function takes two lists as parameters: 'reviews', which contains the reviews, and 'sentiments', which contains the sentiment values associated with each review.

The function iterates through three sentiment categories.

0 for negative  

1 for neutral  

2 for positive  


It then displays examples of reviews corresponding to each category.
For each category, it prints the category number and filters the reviews that have that sentiment value.
Then, it presents the first three examples of reviews from that category.


In [128]:
def examples_reviews_by_sentiments(reviews, sentiments):
    
    for sentiment_value in range(3):
        print(f'For the sentiment analysis category {sentiment_value}, here are some examples of reviwes')
        sentiment_reviews = [reviews for reviews, sentiment in zip(reviews, sentiments) if sentiments == sentiment_value]
        
        for i, reviews in enumerate(sentiment_reviews[:3], start=1):
            print(f'Review {i}: {reviews}')
            
        print('\n')

# 1.3 ETL - australia_user_reviews


The code below loads a dataset from a JSON file, converts it into a pandas DataFrame, and finally returns that DataFrame as df_reviews.

In [129]:
'''Dataset path australia_user_reviews'''

path_review = 'C:\\Users\\migue\\Proyecto_Individual_1_MLOps\\PI MLOps - STEAM\\australian_user_reviews.json'


'''Each line of the dataset is read'''
rows_review = []
with open(path_review, encoding='utf-8') as f:
    '''Each line of the JSON file is iterated over'''
    for line in f.readlines():
        '''Ast.literal_eval is used to evaluate the line as a Python literal expression (converting JSON to Python).'''
        rows_review.append(ast.literal_eval(line))
        
'''It is converted into a DataFrame'''
df_reviews = pd.DataFrame(rows_review)
df_reviews


user_id                                           user_url  \
0      76561197970982479  http://steamcommunity.com/profiles/76561197970...   
1                js41637               http://steamcommunity.com/id/js41637   
2              evcentric             http://steamcommunity.com/id/evcentric   
3                  doctr                 http://steamcommunity.com/id/doctr   
4              maplemage             http://steamcommunity.com/id/maplemage   
...                  ...                                                ...   
25794  76561198306599751  http://steamcommunity.com/profiles/76561198306...   
25795           Ghoustik              http://steamcommunity.com/id/Ghoustik   
25796  76561198310819422  http://steamcommunity.com/profiles/76561198310...   
25797  76561198312638244  http://steamcommunity.com/profiles/76561198312...   
25798        LydiaMorley           http://steamcommunity.com/id/LydiaMorley   

                                                 reviews  
0      [{'funny': '', 'posted': 'Posted November 5, 2...  
1      [{'funny': '', 'posted': 'Posted June 24, 2014...  
2      [{'funny': '', 'posted': 'Posted February 3.',...  
3      [{'funny': '', 'posted': 'Posted October 14, 2...  
4      [{'funny': '3 people found this review funny',...  
...                                                  ...  
25794  [{'funny': '', 'posted': 'Posted May 31.', 'la...  
25795  [{'funny': '', 'posted': 'Posted June 17.', 'l...  
25796  [{'funny': '1 person found this review funny',...  
25797  [{'funny': '', 'posted': 'Posted July 21.', 'l...  
25798  [{'funny': '1 person found this review funny',...  

[25799 rows x 3 columns]

We verify the data types with our function **'check_data_type(df)'**

In [130]:
check_data_type(df_reviews)

field_name         data_type  non_null_%  null_%  nulls
0    user_id   [<class 'str'>]       100.0     0.0      0
1   user_url   [<class 'str'>]       100.0     0.0      0
2    reviews  [<class 'list'>]       100.0     0.0      0

We check for duplicates per column with our function **'check_duplicates_by_columns(df, column)'.**

In [131]:
duplicated_rows = check_duplicates_by_columns(df_reviews, 'user_id')
duplicated_rows

user_id  \
12888               05041129   
5250                05041129   
3133   111222333444555666888   
3134   111222333444555666888   
4139                   29123   
...                      ...   
2721          xXAussieRockXx   
2680             yolofaceguy   
17916            yolofaceguy   
5855               zeroblade   
13975              zeroblade   

                                                user_url  \
12888              http://steamcommunity.com/id/05041129   
5250               http://steamcommunity.com/id/05041129   
3133   http://steamcommunity.com/id/11122233344455566...   
3134   http://steamcommunity.com/id/11122233344455566...   
4139                  http://steamcommunity.com/id/29123   
...                                                  ...   
2721         http://steamcommunity.com/id/xXAussieRockXx   
2680            http://steamcommunity.com/id/yolofaceguy   
17916           http://steamcommunity.com/id/yolofaceguy   
5855              http://steamcommunity.com/id/zeroblade   
13975             http://steamcommunity.com/id/zeroblade   

                                                 reviews  
12888  [{'funny': '', 'posted': 'Posted May 18, 2015....  
5250   [{'funny': '', 'posted': 'Posted May 18, 2015....  
3133   [{'funny': '', 'posted': 'Posted December 22, ...  
3134   [{'funny': '', 'posted': 'Posted December 22, ...  
4139   [{'funny': '', 'posted': 'Posted March 26.', '...  
...                                                  ...  
2721   [{'funny': '', 'posted': 'Posted July 17, 2015...  
2680   [{'funny': '', 'posted': 'Posted October 31, 2...  
17916  [{'funny': '', 'posted': 'Posted October 31, 2...  
5855   [{'funny': '', 'posted': 'Posted November 30, ...  
13975  [{'funny': '', 'posted': 'Posted November 30, ...  

[623 rows x 3 columns]

We've identified 623 duplicated rows in the 'user_id' column, indicating instances where the same identifier ('user_id') appears more than once in DataFrame.  

In addition to detecting duplicates in the 'user_id' column, a more thorough investigation is underway within the nested 'review' structure to determine if the duplication is related to additional information contained in that structure.  

This 'reviw' structure appears to contain comments and other service-related information. The possibility that duplication is due to multiple comments made by the same user is being considered.  

In summary, the analysis aims to identify duplicates in the 'user_id' column while also exploring whether duplication extends to additional information within the nested 'review' structure, especially in cases where a single user may have made multiple comments.

In [132]:
user_id = '05041129'                                                                 # An example user is reviewed
user_reviews = duplicated_rows[duplicated_rows['user_id'] == user_id]['reviews']     # Duplicate rows corresponding to the user are filtered out.

for review_list in user_reviews:                                                     # The list of reviews for the user are iterated over
    for review in review_list:                                                       # Each review in the list is iterated over.
        print(review['review'])                                                      # Print the content of each review.
        
    print('-' * 40)                                                                  # Prints a dividing line after printing all the user's reviews.

This game to me it is so good that it is better than any of the games out their and $15 worth it
this is the best third person game ever that i have played
this will be the  number one game if it have more competitive things
----------------------------------------
This game to me it is so good that it is better than any of the games out their and $15 worth it
this is the best third person game ever that i have played
this will be the  number one game if it have more competitive things
----------------------------------------


The comment in the 'review' is found to be the same, so it is decided to remove the duplicate rows.

In [133]:
df_reviews = df_reviews.drop_duplicates(subset='user_id', keep='first')  # Deletes duplicate rows in the DataFrame df_reviews based on the 'users_id' column.
check_duplicates_by_columns(df_reviews, 'user_id')                       # Call the function to ensure there are no duplicates in the 'users_id' column.

'There are no duplicates'

We analyze the 'reviews' column to see what types of data it holds.

In [134]:
df_reviews['reviews'][0]                    # The data type contained in 'review' is observed.

[{'funny': '',
  'posted': 'Posted November 5, 2011.',
  'last_edited': '',
  'item_id': '1250',
  'helpful': 'No ratings yet',
  'recommend': True,
  'review': 'Simple yet with great replayability. In my opinion does "zombie" hordes and team work better than left 4 dead plus has a global leveling system. Alot of down to earth "zombie" splattering fun for the whole family. Amazed this sort of FPS is so rare.'},
 {'funny': '',
  'posted': 'Posted July 15, 2011.',
  'last_edited': '',
  'item_id': '22200',
  'helpful': 'No ratings yet',
  'recommend': True,
  'review': "It's unique and worth a playthrough."},
 {'funny': '',
  'posted': 'Posted April 21, 2011.',
  'last_edited': '',
  'item_id': '43110',
  'helpful': 'No ratings yet',
  'recommend': True,
  'review': 'Great atmosphere. The gunplay can be a bit chunky at times but at the end of the day this game is definitely worth it and I hope they do a sequel...so buy the game so I get a sequel!'}]

This dataset contains 3 columns and 25799 rows, with no null values.

The columns are:  

**user_id:**        A unique identifier for the user.  
**user_url:**       The URL of the user's profile on streamcommunity.  
**reviews:**        Contains a list of dictionaries. For each user, there is one or more dictionaries with the review. Each dictionary contain:  
    **funny:**          Indicates if someone put a funny emoji on the review.  
    **posted:**         The date the review was posted in the format 'Posted April 21, 20211'.  
    **last_edited:**    The date of the last edit.  
    **item_id:**        The unique identifier of the item, i.e., the game.  
    **helpful:**        The statistic where other users indicate if the information was helpful.  
    **recommend:**      A boolean indicating whether the user recommends the game or not.  
    **review:**         A string sentence with comments about the game.




Each element of the lists is transformed into columns.




In [135]:
df_reviews2 = pd.json_normalize(df_reviews['reviews'])
df_reviews2.head()

0  \
0  {'funny': '', 'posted': 'Posted November 5, 20...   
1  {'funny': '', 'posted': 'Posted June 24, 2014....   
2  {'funny': '', 'posted': 'Posted February 3.', ...   
3  {'funny': '', 'posted': 'Posted October 14, 20...   
4  {'funny': '3 people found this review funny', ...   

                                                   1  \
0  {'funny': '', 'posted': 'Posted July 15, 2011....   
1  {'funny': '', 'posted': 'Posted September 8, 2...   
2  {'funny': '', 'posted': 'Posted December 4, 20...   
3  {'funny': '', 'posted': 'Posted July 28, 2012....   
4  {'funny': '1 person found this review funny', ...   

                                                   2  \
0  {'funny': '', 'posted': 'Posted April 21, 2011...   
1  {'funny': '', 'posted': 'Posted November 29, 2...   
2  {'funny': '', 'posted': 'Posted November 3, 20...   
3  {'funny': '', 'posted': 'Posted June 2, 2012.'...   
4  {'funny': '2 people found this review funny', ...   

                                                   3  \
0                                               None   
1                                               None   
2  {'funny': '', 'posted': 'Posted October 15, 20...   
3  {'funny': '', 'posted': 'Posted June 29, 2014....   
4  {'funny': '', 'posted': 'Posted July 11, 2013....   

                                                   4  \
0                                               None   
1                                               None   
2  {'funny': '', 'posted': 'Posted October 15, 20...   
3  {'funny': '', 'posted': 'Posted November 22, 2...   
4                                               None   

                                                   5     6     7     8     9  
0                                               None  None  None  None  None  
1                                               None  None  None  None  None  
2  {'funny': '', 'posted': 'Posted October 15, 20...  None  None  None  None  
3  {'funny': '', 'posted': 'Posted February 23, 2...  None  None  None  None  
4                                               None  None  None  None  None

We add the 'user_id' and 'user_url' columns, which we had lost, to the separate columns.

In [136]:
df_reviews2 = pd.concat([df_reviews[['user_id', 'user_url']], df_reviews2], axis=1)
df_reviews2.head()

user_id                                           user_url  \
0  76561197970982479  http://steamcommunity.com/profiles/76561197970...   
1            js41637               http://steamcommunity.com/id/js41637   
2          evcentric             http://steamcommunity.com/id/evcentric   
3              doctr                 http://steamcommunity.com/id/doctr   
4          maplemage             http://steamcommunity.com/id/maplemage   

                                                   0  \
0  {'funny': '', 'posted': 'Posted November 5, 20...   
1  {'funny': '', 'posted': 'Posted June 24, 2014....   
2  {'funny': '', 'posted': 'Posted February 3.', ...   
3  {'funny': '', 'posted': 'Posted October 14, 20...   
4  {'funny': '3 people found this review funny', ...   

                                                   1  \
0  {'funny': '', 'posted': 'Posted July 15, 2011....   
1  {'funny': '', 'posted': 'Posted September 8, 2...   
2  {'funny': '', 'posted': 'Posted December 4, 20...   
3  {'funny': '', 'posted': 'Posted July 28, 2012....   
4  {'funny': '1 person found this review funny', ...   

                                                   2  \
0  {'funny': '', 'posted': 'Posted April 21, 2011...   
1  {'funny': '', 'posted': 'Posted November 29, 2...   
2  {'funny': '', 'posted': 'Posted November 3, 20...   
3  {'funny': '', 'posted': 'Posted June 2, 2012.'...   
4  {'funny': '2 people found this review funny', ...   

                                                   3  \
0                                               None   
1                                               None   
2  {'funny': '', 'posted': 'Posted October 15, 20...   
3  {'funny': '', 'posted': 'Posted June 29, 2014....   
4  {'funny': '', 'posted': 'Posted July 11, 2013....   

                                                   4  \
0                                               None   
1                                               None   
2  {'funny': '', 'posted': 'Posted October 15, 20...   
3  {'funny': '', 'posted': 'Posted November 22, 2...   
4                                               None   

                                                   5     6     7     8     9  
0                                               None  None  None  None  None  
1                                               None  None  None  None  None  
2  {'funny': '', 'posted': 'Posted October 15, 20...  None  None  None  None  
3  {'funny': '', 'posted': 'Posted February 23, 2...  None  None  None  None  
4                                               None  None  None  None  None

We use the melt function from Pandas to transform the DataFrame df_reviews2 from a wide to a long format, generating a dictionary for each user. Then, we retain the 'user_id' and 'user_url' columns as unique identifiers.

In [137]:
'''The value columns are specified using list(range(9), and the new column is named 'reviews')'''

df_reviews2 = pd.melt(df_reviews2, id_vars=['user_id', 'user_url'],
                      value_vars=list(range(9)),
                      value_name='reviews')

df_reviews2.head()

user_id                                           user_url  \
0  76561197970982479  http://steamcommunity.com/profiles/76561197970...   
1            js41637               http://steamcommunity.com/id/js41637   
2          evcentric             http://steamcommunity.com/id/evcentric   
3              doctr                 http://steamcommunity.com/id/doctr   
4          maplemage             http://steamcommunity.com/id/maplemage   

  variable                                            reviews  
0        0  {'funny': '', 'posted': 'Posted November 5, 20...  
1        0  {'funny': '', 'posted': 'Posted June 24, 2014....  
2        0  {'funny': '', 'posted': 'Posted February 3.', ...  
3        0  {'funny': '', 'posted': 'Posted October 14, 20...  
4        0  {'funny': '3 people found this review funny', ...

In [138]:
'''There are users who made more reviews than others, and this can be seen with the value None.'''
df_reviews2[df_reviews2['user_id'] == 'js41637']

user_id                              user_url variable  \
1       js41637  http://steamcommunity.com/id/js41637        0   
25800   js41637  http://steamcommunity.com/id/js41637        1   
51599   js41637  http://steamcommunity.com/id/js41637        2   
77398   js41637  http://steamcommunity.com/id/js41637        3   
103197  js41637  http://steamcommunity.com/id/js41637        4   
128996  js41637  http://steamcommunity.com/id/js41637        5   
154795  js41637  http://steamcommunity.com/id/js41637        6   
180594  js41637  http://steamcommunity.com/id/js41637        7   
206393  js41637  http://steamcommunity.com/id/js41637        8   

                                                  reviews  
1       {'funny': '', 'posted': 'Posted June 24, 2014....  
25800   {'funny': '', 'posted': 'Posted September 8, 2...  
51599   {'funny': '', 'posted': 'Posted November 29, 2...  
77398                                                None  
103197                                               None  
128996                                               None  
154795                                               None  
180594                                               None  
206393                                               None

In [139]:
'''We remove the None values in the reviews column.'''
df_reviews2 = df_reviews2.dropna()

'''We verify that only the user_id with the corresponding number of dictionaries remain'''
df_reviews2[df_reviews2['user_id']=='76561197970982479']

user_id                                           user_url  \
0      76561197970982479  http://steamcommunity.com/profiles/76561197970...   
25799  76561197970982479  http://steamcommunity.com/profiles/76561197970...   
51598  76561197970982479  http://steamcommunity.com/profiles/76561197970...   

      variable                                            reviews  
0            0  {'funny': '', 'posted': 'Posted November 5, 20...  
25799        1  {'funny': '', 'posted': 'Posted July 15, 2011....  
51598        2  {'funny': '', 'posted': 'Posted April 21, 2011...

In [140]:
'''We split each key of the reviews column into separate columns.'''
df_reviews = df_reviews2['reviews'].apply(pd.Series, dtype='object')
df_reviews = df_reviews.add_prefix('reviews_')
df_reviews.head()

reviews_funny            reviews_posted  \
0                                    Posted November 5, 2011.   
1                                       Posted June 24, 2014.   
2                                          Posted February 3.   
3                                    Posted October 14, 2013.   
4  3 people found this review funny    Posted April 15, 2014.   

  reviews_last_edited reviews_item_id  \
0                                1250   
1                              251610   
2                              248820   
3                              250320   
4                              211420   

                                   reviews_helpful  reviews_recommend  \
0                                   No ratings yet               True   
1  15 of 20 people (75%) found this review helpful               True   
2                                   No ratings yet               True   
3   2 of 2 people (100%) found this review helpful               True   
4  35 of 43 people (81%) found this review helpful               True   

                                      reviews_review  
0  Simple yet with great replayability. In my opi...  
1  I know what you think when you see this title ...  
2  A suitably punishing roguelike platformer.  Wi...  
3  This game... is so fun. The fight sequences ha...  
4                                            Git gud

In [141]:
'''It is merged with the 'user_id' and 'user_url'.'''
df_reviews = pd.concat([df_reviews2[['user_id', 'user_url']], df_reviews], axis=1)
df_reviews.head()

user_id                                           user_url  \
0  76561197970982479  http://steamcommunity.com/profiles/76561197970...   
1            js41637               http://steamcommunity.com/id/js41637   
2          evcentric             http://steamcommunity.com/id/evcentric   
3              doctr                 http://steamcommunity.com/id/doctr   
4          maplemage             http://steamcommunity.com/id/maplemage   

                      reviews_funny            reviews_posted  \
0                                    Posted November 5, 2011.   
1                                       Posted June 24, 2014.   
2                                          Posted February 3.   
3                                    Posted October 14, 2013.   
4  3 people found this review funny    Posted April 15, 2014.   

  reviews_last_edited reviews_item_id  \
0                                1250   
1                              251610   
2                              248820   
3                              250320   
4                              211420   

                                   reviews_helpful  reviews_recommend  \
0                                   No ratings yet               True   
1  15 of 20 people (75%) found this review helpful               True   
2                                   No ratings yet               True   
3   2 of 2 people (100%) found this review helpful               True   
4  35 of 43 people (81%) found this review helpful               True   

                                      reviews_review  
0  Simple yet with great replayability. In my opi...  
1  I know what you think when you see this title ...  
2  A suitably punishing roguelike platformer.  Wi...  
3  This game... is so fun. The fight sequences ha...  
4                                            Git gud

We replace missing values with null values.

In [142]:
df_reviews.replace('', None, inplace=True)
df_reviews.head()

user_id                                           user_url  \
0  76561197970982479  http://steamcommunity.com/profiles/76561197970...   
1            js41637               http://steamcommunity.com/id/js41637   
2          evcentric             http://steamcommunity.com/id/evcentric   
3              doctr                 http://steamcommunity.com/id/doctr   
4          maplemage             http://steamcommunity.com/id/maplemage   

                      reviews_funny            reviews_posted  \
0                              None  Posted November 5, 2011.   
1                              None     Posted June 24, 2014.   
2                              None        Posted February 3.   
3                              None  Posted October 14, 2013.   
4  3 people found this review funny    Posted April 15, 2014.   

  reviews_last_edited reviews_item_id  \
0                None            1250   
1                None          251610   
2                None          248820   
3                None          250320   
4                None          211420   

                                   reviews_helpful  reviews_recommend  \
0                                   No ratings yet               True   
1  15 of 20 people (75%) found this review helpful               True   
2                                   No ratings yet               True   
3   2 of 2 people (100%) found this review helpful               True   
4  35 of 43 people (81%) found this review helpful               True   

                                      reviews_review  
0  Simple yet with great replayability. In my opi...  
1  I know what you think when you see this title ...  
2  A suitably punishing roguelike platformer.  Wi...  
3  This game... is so fun. The fight sequences ha...  
4                                            Git gud

We verify the data after unnesting.

In [143]:
check_data_type(df_reviews)

field_name                            data_type  non_null_%  \
0              user_id                      [<class 'str'>]      100.00   
1             user_url                      [<class 'str'>]      100.00   
2        reviews_funny  [<class 'NoneType'>, <class 'str'>]       13.76   
3       reviews_posted                      [<class 'str'>]      100.00   
4  reviews_last_edited  [<class 'NoneType'>, <class 'str'>]       10.28   
5      reviews_item_id                      [<class 'str'>]      100.00   
6      reviews_helpful                      [<class 'str'>]      100.00   
7    reviews_recommend                     [<class 'bool'>]      100.00   
8       reviews_review  [<class 'str'>, <class 'NoneType'>]       99.95   

   null_%  nulls  
0    0.00      0  
1    0.00      0  
2   86.24  49498  
3    0.00      0  
4   89.72  51499  
5    0.00      0  
6    0.00      0  
7    0.00      0  
8    0.05     30

Eliminate columns with more missing data, which are 'review_funny' (index 2) and 'reviews_last_edited' (index 4).

In [144]:
df_reviews = df_reviews.drop(columns = ['reviews_funny', 'reviews_last_edited'])
df_reviews.columns

Index(['user_id', 'user_url', 'reviews_posted', 'reviews_item_id',
       'reviews_helpful', 'reviews_recommend', 'reviews_review'],
      dtype='object')

We transform the data into YYYY-MM-DD format using our convert_date function.

In [145]:
df_reviews['reviews_date'] = df_reviews['reviews_posted'].apply(convert_date)
df_reviews['reviews_date']

0               11-11-05
1               14-06-24
2         Invalid format
3               13-10-14
4               14-04-15
               ...      
231291          14-08-15
231293          14-08-02
231419          15-07-31
231499          15-12-20
231501    Invalid format
Name: reviews_date, Length: 57397, dtype: object

There are some records that will remain in an invalid format.

In [146]:
df_reviews[df_reviews['reviews_date'] == 'Invalid format']

user_id                                           user_url  \
2               evcentric             http://steamcommunity.com/id/evcentric   
6       76561198079601835  http://steamcommunity.com/profiles/76561198079...   
7             MeaTCompany           http://steamcommunity.com/id/MeaTCompany   
9       76561198156664158  http://steamcommunity.com/profiles/76561198156...   
10      76561198077246154  http://steamcommunity.com/profiles/76561198077...   
...                   ...                                                ...   
223569  76561198040184950  http://steamcommunity.com/profiles/76561198040...   
226105  76561198046474248  http://steamcommunity.com/profiles/76561198046...   
228109         dmitry_who            http://steamcommunity.com/id/dmitry_who   
229231  76561198079507136  http://steamcommunity.com/profiles/76561198079...   
231501       EchoXSilence          http://steamcommunity.com/id/EchoXSilence   

            reviews_posted reviews_item_id  \
2       Posted February 3.          248820   
6           Posted May 20.             730   
7          Posted July 24.             730   
9          Posted June 16.          252950   
10         Posted June 11.             440   
...                    ...             ...   
223569    Posted April 12.          394690   
226105    Posted March 28.          234140   
228109      Posted May 17.          376210   
229231   Posted January 3.             730   
231501  Posted February 3.             570   

                                        reviews_helpful  reviews_recommend  \
2                                        No ratings yet               True   
6          0 of 1 people (0%) found this review helpful               True   
7                                        No ratings yet               True   
9          0 of 1 people (0%) found this review helpful               True   
10                                       No ratings yet               True   
...                                                 ...                ...   
223569                                   No ratings yet               True   
226105                                   No ratings yet               True   
228109  10 of 28 people (36%) found this review helpful               True   
229231                                   No ratings yet              False   
231501     0 of 2 people (0%) found this review helpful              False   

                                           reviews_review    reviews_date  
2       A suitably punishing roguelike platformer.  Wi...  Invalid format  
6                                           ZIKA DO BAILE  Invalid format  
7                           BEST GAME IN THE BLOODY WORLD  Invalid format  
9                                                 love it  Invalid format  
10                                                 mt bom  Invalid format  
...                                                   ...             ...  
223569  I cannot say much right now due to the game no...  Invalid format  
226105  Oh what a day .., What a lovely day to play th...  Invalid format  
228109  ░░░░░░░░░░░█▀▀░░█░░░░░░░░░░░▄▀▀▀▀░░░░░█▄▄░░░░░...  Invalid format  
229231                                          got VACed  Invalid format  
231501  DOTA 2.normal skill - 2ez for me>high skill - ...  Invalid format  

[9771 rows x 8 columns]

We are removing the 'reviews_posted' column due to insufficient information, as pfrases like " Posted March 28.	" do not contribute to our analysis.

In [147]:
df_reviews = df_reviews.drop('reviews_posted', axis=1)
df_reviews.columns

Index(['user_id', 'user_url', 'reviews_item_id', 'reviews_helpful',
       'reviews_recommend', 'reviews_review', 'reviews_date'],
      dtype='object')

We examine the data in the 'reviews_review'  column and remove any null values.

In [148]:
df_reviews = df_reviews.dropna(subset=('reviews_review'))
check_data_type(df_reviews)

field_name         data_type  non_null_%  null_%  nulls
0            user_id   [<class 'str'>]       100.0     0.0      0
1           user_url   [<class 'str'>]       100.0     0.0      0
2    reviews_item_id   [<class 'str'>]       100.0     0.0      0
3    reviews_helpful   [<class 'str'>]       100.0     0.0      0
4  reviews_recommend  [<class 'bool'>]       100.0     0.0      0
5     reviews_review   [<class 'str'>]       100.0     0.0      0
6       reviews_date   [<class 'str'>]       100.0     0.0      0

In [149]:
df_reviews.head()

user_id                                           user_url  \
0  76561197970982479  http://steamcommunity.com/profiles/76561197970...   
1            js41637               http://steamcommunity.com/id/js41637   
2          evcentric             http://steamcommunity.com/id/evcentric   
3              doctr                 http://steamcommunity.com/id/doctr   
4          maplemage             http://steamcommunity.com/id/maplemage   

  reviews_item_id                                  reviews_helpful  \
0            1250                                   No ratings yet   
1          251610  15 of 20 people (75%) found this review helpful   
2          248820                                   No ratings yet   
3          250320   2 of 2 people (100%) found this review helpful   
4          211420  35 of 43 people (81%) found this review helpful   

   reviews_recommend                                     reviews_review  \
0               True  Simple yet with great replayability. In my opi...   
1               True  I know what you think when you see this title ...   
2               True  A suitably punishing roguelike platformer.  Wi...   
3               True  This game... is so fun. The fight sequences ha...   
4               True                                            Git gud   

     reviews_date  
0        11-11-05  
1        14-06-24  
2  Invalid format  
3        13-10-14  
4        14-04-15

C:\Users\migue\Proyecto_Individual_1_MLOps\PI MLOps - STEAM\clean datasets

We save the dataset as Australian_user_reviews_clean.csv in the clean datasets folder.

In [151]:
'''Especify the directory path and the CSV file name.'''

directory = 'C:\\Users\\migue\\Proyecto_Individual_1_MLOps\\PI MLOps - STEAM\\clean datasets'
file_name = 'Australian_user_reviews_clean.csv'
full_path = f'{directory}/{file_name}'


df_reviews.to_csv(full_path, index=False)

print(f'The file was successfully saved')

The file was successfully saved
